code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/BaseConstraintsProvider.java

```diff
@@ -106,49 +110,59 @@ protected void addComputeConstraints(AllocationAttempt attempt, List<Constraint>
     protected void addStorageConstraints(AllocationAttempt attempt, List<Constraint> constraints) {
         for (Map.Entry<Volume, Set<StoragePool>> entry : attempt.getPools().entrySet()) {
             Volume volume = entry.getKey();
-            List<? extends VolumeStoragePoolMap> vspms = mapDao.findNonRemoved(VolumeStoragePoolMap.class, Volume.class, volume.getId());
-            boolean alreadyMappedToPool = vspms.size() > 0;
+            boolean alreadyMappedToPool = entry.getValue().size() > 0;
             VolumeValidStoragePoolConstraint volumeToPoolConstraint = new VolumeValidStoragePoolConstraint(volume, alreadyMappedToPool);
-            if (alreadyMappedToPool) {
-                for (VolumeStoragePoolMap vspm : vspms) {
-                    volumeToPoolConstraint.getStoragePools().add(vspm.getStoragePoolId());
-                    ValidHostsConstraint hostSet = new ValidHostsConstraint();
-                    for (StoragePoolHostMap sphm : mapDao.findNonRemoved(StoragePoolHostMap.class, StoragePool.class, vspm.getStoragePoolId())) {
-                        hostSet.addHost(sphm.getHostId());
+            for (StoragePool pool : entry.getValue()) {
+                volumeToPoolConstraint.getStoragePools().add(pool.getId());
+                storagePoolToHostConstraint(constraints, pool);
+            }
+
+            if (!alreadyMappedToPool) {
+                String driver = DataAccessor.fieldString(volume, VolumeConstants.FIELD_VOLUME_DRIVER);
+                boolean restrictToUnmanagedPool = true;
+                if (StringUtils.isNotEmpty(driver) && !VolumeConstants.LOCAL_DRIVER.equals(driver)) {
+                    StoragePool pool = storagePoolDao.findStoragePoolByDriverName(volume.getAccountId(), driver);
```

Comment:
@alena1108 I don't think it needs logged because not finding a pool is  a normal use case. Users can put whatever they want in the volumeDriver field and if it doesn't map to storage pool, it just means that they want to use a volume driver that we are not orchestrating.

For example, if they've already deployed flocker on the hosts as a docker plugin, they would put flocker as the volume driver and that would be just fine.